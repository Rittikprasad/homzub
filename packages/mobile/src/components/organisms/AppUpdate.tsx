/**
 * Handling Force Update, Maintenance logic using remote config
 * Doc for ref: https://rnfirebase.io/remote-config/usage
 */

import React, { useEffect, useState } from 'react';
import { Linking, StyleSheet, View } from 'react-native';
import remoteConfig from '@react-native-firebase/remote-config';
import { AppModes, ConfigHelper, ConfigModes } from '@homzhub/common/src/utils/ConfigHelper';
import { DeviceUtils } from '@homzhub/common/src/utils/DeviceUtils';
import { Logger } from '@homzhub/common/src/utils/Logger';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { I18nService } from '@homzhub/common/src/services/Localization/i18nextService';
import Rocket from '@homzhub/common/src/assets/images/rocket.svg';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';

// Remote config Parameters
enum ConfigKeys {
  MINIMUM_VERSION = 'minimum_version',
  LATEST_VERSION = 'latest_version',
  IS_UNDER_MAINTENANCE = 'is_under_maintenance',
  MESSAGES = 'messages',
}

interface IMessages {
  update: string;
  maintenance: string;
}

interface IConfig {
  [key: string]: any;
}

interface IOwnState {
  isForceUpdate: boolean;
  isUpdateAvailable: boolean;
  isUnderMaintenance: boolean;
  messages: IMessages;
}

const initialState: IOwnState = {
  isForceUpdate: false,
  isUpdateAvailable: false,
  isUnderMaintenance: false,
  messages: {
    update: '',
    maintenance: '',
  },
};

const AppUpdate = (): React.ReactElement => {
  const [configData, setConfigData] = useState<IConfig[]>([]);
  const [configValues, setConfigValues] = useState<IOwnState>(initialState);
  const [isSheetVisible, setVisibility] = useState(false);

  useEffect(() => {
    remoteConfig()
      .setConfigSettings({
        minimumFetchIntervalMillis: 30000, // Global Caching frequency
      })
      .then(() => {
        // Set Default before fetching config from firebase
        remoteConfig()
          .setDefaults({
            is_under_maintenance: false,
            latest_version: '',
            minimum_version: '',
          })
          .then(() => remoteConfig().fetchAndActivate()) // Fetch config and Activate
          .then(() => {
            const config = remoteConfig().getAll(); // Get All config list

            if (config) {
              const formattedData = Object.entries(config).map((item) => {
                const [key, entry] = item as any;
                return {
                  [key]: JSON.parse(entry.asString()),
                };
              });
              setConfigData(formattedData);
            }
          })
          .then(() => remoteConfig().fetch()); // Fetching data for next process
      })

      .catch((e) => {
        Logger.warn(e.message);
      });
  }, []);

  useEffect(() => {
    const updatedData = { ...configValues };
    const configMode = ConfigHelper.getConfigMode();
    const appMode = (configMode === ConfigModes.PROD ? AppModes.RELEASE : AppModes.DEBUG).toLowerCase();

    // Structuring config and maintaining local state to handle UI
    configData.forEach((item) => {
      if (Object.keys(item)[0] === ConfigKeys.MINIMUM_VERSION) {
        updatedData.isForceUpdate = item[ConfigKeys.MINIMUM_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.LATEST_VERSION) {
        updatedData.isUpdateAvailable = item[ConfigKeys.LATEST_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.IS_UNDER_MAINTENANCE) {
        updatedData.isUnderMaintenance = item[ConfigKeys.IS_UNDER_MAINTENANCE][appMode];
      } else {
        updatedData.messages = item[ConfigKeys.MESSAGES];
      }
    });

    const isVisible = updatedData.isUpdateAvailable || updatedData.isUnderMaintenance || updatedData.isForceUpdate;

    setVisibility(isVisible);
    setConfigValues(updatedData);
  }, [configData]);

  const onUpdateNow = async (): Promise<void> => {
    const { getAppleStoreUrl, getGooglePlayStoreUrl } = ConfigHelper;
    await Linking.openURL(PlatformUtils.isIOS() ? getAppleStoreUrl() : getGooglePlayStoreUrl());
  };

  const { isForceUpdate, messages, isUnderMaintenance, isUpdateAvailable } = configValues;
  const isAlwaysVisible = isForceUpdate || isUnderMaintenance;
  return (
    <BottomSheet
      visible={isSheetVisible}
      isShadowView
      headerTitle={isUpdateAvailable ? I18nService.t('common:timeToUpdate') : I18nService.t('common:underMaintenance')}
      sheetHeight={isUpdateAvailable ? 420 : 300}
      isCloseOnDrag={!isAlwaysVisible}
      isAlwaysVisible={isAlwaysVisible}
      onCloseSheet={(): void => setVisibility(false)}
    >
      <View style={styles.container}>
        <Rocket style={styles.iconStyle} />
        <Label type="large" style={styles.message}>
          {isUpdateAvailable ? messages?.update : messages?.maintenance}
        </Label>
        {!isUnderMaintenance && (
          <Button
            type="primary"
            title={I18nService.t('common:updateNow')}
            containerStyle={styles.updateButton}
            onPress={onUpdateNow}
          />
        )}
        {!isAlwaysVisible && (
          <Button
            type="secondaryOutline"
            title={I18nService.t('common:doLater')}
            containerStyle={styles.secondaryButton}
            textStyle={styles.buttonTitle}
            onPress={(): void => setVisibility(false)}
          />
        )}
      </View>
    </BottomSheet>
  );
};

export default AppUpdate;

const styles = StyleSheet.create({
  container: {
    padding: 32,
    flex: 1,
  },
  iconStyle: {
    alignSelf: 'center',
  },
  message: {
    textAlign: 'center',
    marginVertical: 16,
    color: theme.colors.darkTint3,
  },
  updateButton: {
    flex: 0,
    height: 46,
    marginVertical: 10,
  },
  secondaryButton: {
    height: 46,
    flex: 0,
  },
  buttonTitle: {
    color: theme.colors.darkTint3,
  },
});
