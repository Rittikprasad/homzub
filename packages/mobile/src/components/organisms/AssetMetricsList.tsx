import React, { useState, useCallback } from 'react';
import { StyleProp, StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useRoute } from '@react-navigation/native';
import { useNavigation } from '@react-navigation/core';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import Logo from '@homzhub/common/src/assets/images/homzhubDashboard.svg';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { PaginationComponent } from '@homzhub/mobile/src/components/atoms/PaginationComponent';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
import { AssetMetrics } from '@homzhub/mobile/src/components/molecules/AssetMetrics';

export interface IMetricsData {
  name: string;
  count: string | number;
  label?: string;
  id?: number;
  isCurrency?: boolean;
  colorCode: string;
  code?: string;
}

interface IProps {
  data: IMetricsData[];
  title?: string;
  subscription?: string;
  selectedAssetType?: string;
  numOfElements?: number;
  isSubTextRequired?: boolean;
  onPlusIconClicked?: () => void;
  onMetricsClicked?: (name: string) => void;
  containerStyle?: StyleProp<ViewStyle>;
  subTitleText?: string;
  headerIcon?: string;
  showBackIcon?: boolean;
}

const COMPONENT_PADDING = 12;
const SLIDER_WIDTH = theme.viewport.width - theme.layout.screenPadding * 2;
const AssetMetricsList = (props: IProps): React.ReactElement => {
  const {
    title = 0,
    data,
    selectedAssetType,
    onPlusIconClicked,
    containerStyle,
    onMetricsClicked,
    numOfElements = 3,
    isSubTextRequired = true,
    subTitleText,
    headerIcon,
    showBackIcon = false,
  } = props;

  // HOOKS
  const { t } = useTranslation();
  const { goBack, setParams } = useNavigation();
  const { params } = useRoute();
  const [activeIndex, setActiveIndex] = useState(0);
  // HELPERS
  const bubblePlusIcon = useCallback((): void => {
    if (onPlusIconClicked) {
      onPlusIconClicked();
    }
  }, [onPlusIconClicked]);

  const splitData = useCallback((): IMetricsData[][] => {
    const newArr = [];

    for (let i = 0; i < data.length; i += numOfElements) {
      newArr.push(data.slice(i, i + numOfElements));
    }

    return newArr;
  }, [data, numOfElements]);

  const onGoBack = (): void => {
    goBack();
    // @ts-ignore
    if (params && params?.isFromNavigation) {
      setParams({ isFromNavigation: false });
    }
  };
  // HELPERS END

  const renderItem = useCallback(
    (items: IMetricsData[]): React.ReactElement => {
      return (
        <View style={styles.sliderRow}>
          {items.map((item: IMetricsData) => {
            const cardStyle = {
              minWidth: (SLIDER_WIDTH - COMPONENT_PADDING * (numOfElements + 1)) / numOfElements,
            };
            // TODO: (Shikha) - Connect with BE to handle all logic with CODE
            const handlePress = (): void => onMetricsClicked && onMetricsClicked(item.code || item.name);
            return (
              <AssetMetrics
                key={`${item.label ?? item.name}`}
                header={item.label ?? item.name}
                value={item.count}
                isCurrency={item.isCurrency ?? false}
                cardStyle={cardStyle}
                colorCode={item.colorCode}
                onPressMetrics={handlePress}
                selectedAssetType={selectedAssetType}
                // Todo : Check in multiple devices
                maxLength={items.length === 2 ? 15 : undefined}
              />
            );
          })}
        </View>
      );
    },
    [onMetricsClicked, numOfElements, selectedAssetType]
  );

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={[styles.header, !isSubTextRequired && styles.financialView]}>
        {showBackIcon && (
          <TouchableOpacity onPress={onGoBack} style={styles.backIcon}>
            <Icon name={icons.leftArrow} size={25} color={theme.colors.blue} />
          </TouchableOpacity>
        )}
        {headerIcon ? (
          <View>
            <Icon name={headerIcon} size={30} style={styles.icon} color={theme.colors.blue} />
            <View style={[styles.headerIcon, styles.logo]} />
          </View>
        ) : (
          <Logo style={styles.logo} width={50} height={50} />
        )}
        <View style={styles.headerCenter}>
          <Text type="regular" textType="bold" style={styles.assetCount}>
            {title}
          </Text>
          <View style={styles.propertiesRow}>
            {isSubTextRequired && (
              <Label type="large" textType="semiBold" style={styles.propertyText}>
                {subTitleText || t('common:properties')}
              </Label>
            )}
          </View>
        </View>
        {onPlusIconClicked && (
          <TouchableOpacity onPress={bubblePlusIcon} style={styles.plusIconContainer}>
            <Icon name={icons.circularPlus} size={32} color={theme.colors.primaryColor} testID="icnPlus" />
          </TouchableOpacity>
        )}
      </View>
      <SnapCarousel
        carouselData={splitData()}
        carouselItem={renderItem}
        activeIndex={activeIndex}
        onSnapToItem={setActiveIndex}
        sliderWidth={SLIDER_WIDTH}
        itemWidth={SLIDER_WIDTH}
      />
      <PaginationComponent
        dotsLength={splitData().length}
        activeSlide={activeIndex}
        containerStyle={styles.paginationContainer}
        activeDotStyle={[styles.dot, styles.activeDot]}
        inactiveDotStyle={[styles.dot, styles.inactiveDot]}
      />
    </View>
  );
};

const memoizedComponent = React.memo(AssetMetricsList);
export { memoizedComponent as AssetMetricsList };

const styles = StyleSheet.create({
  container: {
    paddingVertical: 12,
    borderRadius: 4,
    backgroundColor: theme.colors.white,
  },
  header: {
    marginHorizontal: 12,
    flexDirection: 'row',
  },
  headerCenter: {
    flex: 1,
  },
  propertiesRow: {
    flexDirection: 'row',
  },
  sliderRow: {
    flexDirection: 'row',
  },
  paginationContainer: {
    paddingVertical: 0,
    marginTop: 12,
  },
  assetCount: {
    color: theme.colors.darkTint1,
  },
  propertyText: {
    color: theme.colors.darkTint4,
  },
  logo: {
    marginEnd: 12,
  },
  dot: {
    width: 8.5,
    height: 8.5,
  },
  activeDot: {
    borderWidth: 1.5,
  },
  inactiveDot: {
    backgroundColor: theme.colors.disabled,
    borderWidth: 0,
  },
  financialView: {
    alignItems: 'center',
  },
  headerIcon: {
    backgroundColor: theme.colors.blue,
    width: 50,
    height: 50,
    opacity: 0.1,
    borderRadius: 30,
  },
  icon: {
    position: 'absolute',
    left: 10,
    top: 10,
  },
  plusIconContainer: {
    justifyContent: 'center',
    padding: 5,
  },
  backIcon: {
    alignSelf: 'center',
    marginRight: 10,
  },
});
