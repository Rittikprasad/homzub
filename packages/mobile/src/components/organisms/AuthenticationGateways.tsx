import React from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { WithTranslation, withTranslation } from 'react-i18next';
import { StackNavigationProp } from '@react-navigation/stack';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { SocialAuthUtils } from '@homzhub/common/src/utils/SocialAuthUtils';
import { AuthService } from '@homzhub/mobile/src/services/AuthService';
import { ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import Apple from '@homzhub/common/src/assets/images/apple.svg';
import Google from '@homzhub/common/src/assets/images/google.svg';
import Facebook from '@homzhub/common/src/assets/images/facebook.svg';
import LinkedIn from '@homzhub/common/src/assets/images/linkedin.svg';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import CommonAuthenticationHoc from '@homzhub/common/src/components/organisms/CommonAuthenticationHOC';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { SocialAuthProvider } from '@homzhub/common/src/domain/models/SocialAuthProvider';
import { ISocialUserData, SocialAuthKeys } from '@homzhub/common/src/constants/SocialAuthProviders';
import { IUserTokens } from '@homzhub/common/src/services/storage/StorageService';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IDispatchProps {
  loginSuccess: (data: IUserTokens) => void;
}

interface IAuthenticationGatewayProps {
  isFromLogin: boolean;
  onEmailLogin?: () => void;
  testID?: string;
  navigation?: StackNavigationProp<AuthStackParamList, ScreensKeys.SignUp | ScreensKeys.Login>;
  toggleLoading?: (isLoading: boolean) => void;
  onSuccessCallback?: () => void;
  isSocialAllowed?: boolean;
}

type Props = IAuthenticationGatewayProps & WithTranslation & IDispatchProps;

interface IOwnState {
  authProviders: SocialAuthProvider[];
}

class AuthenticationGateways extends React.PureComponent<Props, IOwnState> {
  public state = {
    authProviders: [],
  };

  public componentDidMount(): void {
    SocialAuthUtils.fetchSocialMedia((response) => {
      if (PlatformUtils.isIOS()) {
        response.unshift(
          ObjectMapper.deserialize(SocialAuthProvider, {
            provider: SocialAuthKeys.Apple,
            description: 'Apple',
            visible: PlatformUtils.isIOS(),
          })
        );
      }

      this.setState({ authProviders: response });
    });
  }

  public render(): React.ReactNode {
    const { onEmailLogin, isFromLogin, isSocialAllowed = true } = this.props;

    return (
      <CommonAuthenticationHoc isFromLogin={isFromLogin} onEmailLogin={onEmailLogin}>
        {!PlatformUtils.isIOS() && isSocialAllowed && this.renderSocialAuths()}
      </CommonAuthenticationHoc>
    );
  }

  private renderSocialAuths = (): React.ReactNode => {
    const { authProviders } = this.state;

    return authProviders.map((auth: SocialAuthProvider) => {
      const initiateSocialLogin = async (): Promise<void> => {
        await this.initiateSocialAuthentication(auth.provider);
      };

      return (
        <>
          {auth.visible && (
            <TouchableOpacity style={styles.alignToCenter} onPress={initiateSocialLogin} key={auth.provider}>
              {this.getImageUrlForLib(auth.provider)}
              <Label type="regular" textType="regular" style={styles.iconTextStyle}>
                {auth.description}
              </Label>
            </TouchableOpacity>
          )}
        </>
      );
    });
  };

  private onSocialAuthSuccess = async (userData: ISocialUserData): Promise<void> => {
    const { loginSuccess, navigation, isFromLogin, onSuccessCallback } = this.props;

    if (navigation) {
      await SocialAuthUtils.onSocialAuthSuccess(
        userData,
        () => {
          navigation.navigate(ScreensKeys.MobileVerification, {
            isFromLogin,
            userData,
          });
        },
        (response) => {
          loginSuccess(response);
          if (onSuccessCallback) {
            onSuccessCallback();
          }
        }
      );
    }
  };

  private initiateSocialAuthentication = async (key: string): Promise<void> => {
    const { isFromLogin, navigation } = this.props;
    this.setLoading(true);
    if (key === SocialAuthKeys.Google) {
      await AuthService.signInWithGoogle(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Apple) {
      await AuthService.signInWithApple(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Facebook) {
      await AuthService.signInWithFacebook(this.onSocialAuthSuccess, isFromLogin && navigation);
      this.setLoading(false);
    }
  };

  private setLoading = (isLoading: boolean): void => {
    const { toggleLoading } = this.props;
    if (toggleLoading) {
      toggleLoading(isLoading);
    }
  };

  private getImageUrlForLib = (key: string): React.ReactNode => {
    switch (key) {
      case SocialAuthKeys.Facebook:
        return <Facebook />;
      case SocialAuthKeys.Google:
        return <Google />;
      case SocialAuthKeys.LinkedIn:
        return <LinkedIn />;
      case SocialAuthKeys.Apple:
        return <Apple />;
      default:
        return null;
    }
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { loginSuccess } = UserActions;
  return bindActionCreators(
    {
      loginSuccess,
    },
    dispatch
  );
};

const HOC = connect(
  null,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.auth)(AuthenticationGateways));
export { HOC as AuthenticationGateways };

const styles = StyleSheet.create({
  iconTextStyle: {
    marginTop: 4,
  },
  alignToCenter: {
    alignItems: 'center',
  },
});
