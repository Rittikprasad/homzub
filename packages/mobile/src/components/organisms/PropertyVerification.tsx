import React from 'react';
import { StyleSheet } from 'react-native';
import DocumentPicker from 'react-native-document-picker';
import ImagePicker from 'react-native-image-crop-picker';
import { WithTranslation, withTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { IDocsProps, ListingService } from '@homzhub/common/src/services/Property/ListingService';
import { AttachmentService } from '@homzhub/common/src/services/AttachmentService';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import VerificationTypes from '@homzhub/common/src/components/organisms/VerificationTypes';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { AllowedAttachmentFormats } from '@homzhub/common/src/domain/models/Attachment';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import {
  ExistingVerificationDocuments,
  IExistingVerificationDocuments,
  IPostVerificationDocuments,
  VerificationDocumentCategory,
  VerificationDocumentTypes,
} from '@homzhub/common/src/domain/models/VerificationDocuments';
import { IUpdateAssetParams } from '@homzhub/common/src/domain/repositories/interfaces';
import { AttachmentError, AttachmentType } from '@homzhub/common/src/constants/AttachmentTypes';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IPropertyVerificationState {
  verificationTypes: VerificationDocumentTypes[];
  existingDocuments: ExistingVerificationDocuments[];
  localDocuments: ExistingVerificationDocuments[];
  isLoading: boolean;
}

interface IProps {
  typeOfPlan: TypeOfPlan;
  updateStep: () => void;
  propertyId: number;
  lastVisitedStep: ILastVisitedStep;
}

type Props = WithTranslation & IProps;

export class PropertyVerification extends React.PureComponent<Props, IPropertyVerificationState> {
  public state = {
    verificationTypes: [],
    existingDocuments: [],
    localDocuments: [],
    isLoading: false,
  };

  public componentDidMount = async (): Promise<void> => {
    const { propertyId } = this.props;
    await ListingService.getExistingDocuments(propertyId, this.updateState);
  };

  public render(): React.ReactElement {
    const { t, typeOfPlan } = this.props;
    const { existingDocuments, localDocuments, isLoading, verificationTypes } = this.state;
    const totalDocuments = existingDocuments.concat(localDocuments);

    const uploadedTypes = totalDocuments.map((doc: ExistingVerificationDocuments) => doc.verificationDocumentType.name);
    const containsAllReqd = uploadedTypes.length === verificationTypes.length;

    return (
      <>
        <VerificationTypes
          typeOfPlan={typeOfPlan}
          existingDocuments={existingDocuments}
          localDocuments={localDocuments}
          handleUpload={this.handleVerificationDocumentUploads}
          deleteDocument={this.onDeleteDocument}
          handleTypes={this.handleVerificationTypes}
        />
        <Button
          type="primary"
          title={t('common:continue')}
          disabled={!containsAllReqd || isLoading}
          containerStyle={styles.buttonStyle}
          onPress={this.postPropertyVerificationDocuments}
        />
      </>
    );
  }

  // HANDLERS START

  public onDeleteDocument = async (
    document: ExistingVerificationDocuments,
    isLocalDocument?: boolean
  ): Promise<void> => {
    const { localDocuments, existingDocuments } = this.state;
    const { propertyId } = this.props;
    await ListingService.deleteDocument(
      {
        document,
        localDocuments,
        existingDocuments,
        isLocalDocument,
        propertyId,
      },
      this.updateState
    );
  };

  public handleVerificationTypes = (types: VerificationDocumentTypes[]): void => {
    this.setState({
      verificationTypes: types,
    });
  };

  public handleVerificationDocumentUploads = async (data: VerificationDocumentTypes): Promise<void> => {
    const verificationDocumentId = data.id;
    const verificationDocumentType = data.name;
    if (verificationDocumentType === VerificationDocumentCategory.SELFIE_ID_PROOF) {
      this.captureSelfie(verificationDocumentId, data);
    } else {
      await this.uploadDocument(verificationDocumentId, data);
    }
  };

  public captureSelfie = (verificationDocumentId: number, data: VerificationDocumentTypes): void => {
    ImagePicker.openCamera({
      width: 400,
      height: 400,
      compressImageMaxWidth: 400,
      compressImageMaxHeight: 400,
      compressImageQuality: PlatformUtils.isAndroid() ? 1 : 0.8,
      useFrontCamera: true,
      cropping: true,
    })
      .then((image: any) => {
        const source = {
          uri: image.path,
          type: image.mime,
          name: PlatformUtils.isIOS()
            ? image.filename ?? image.path.substring(image.path.lastIndexOf('/') + 1)
            : image.path.substring(image.path.lastIndexOf('/') + 1),
        };
        this.updateLocalDocuments(verificationDocumentId, source, data);
      })
      .catch((e) => {
        if (e.code !== 'E_PICKER_CANCELLED') {
          AlertHelper.error({ message: e.message });
        }
      });
  };

  public uploadDocument = async (verificationDocumentId: number, data: VerificationDocumentTypes): Promise<void> => {
    try {
      const document = await DocumentPicker.pick({
        type: [DocumentPicker.types.allFiles],
      });
      if (Object.values(AllowedAttachmentFormats).includes(document.type)) {
        const source = { uri: document.uri, type: document.type, name: document.name };
        this.updateLocalDocuments(verificationDocumentId, source, data);
      } else {
        AlertHelper.error({ message: data.helpText });
      }
    } catch (e) {
      if (!DocumentPicker.isCancel(e)) {
        AlertHelper.error({ message: e.message });
      }
    }
  };

  public updateLocalDocuments = (
    verificationDocumentId: number,
    source: { uri: string; type: string; name: string },
    data: VerificationDocumentTypes
  ): void => {
    const imageObject: IExistingVerificationDocuments = {
      id: null,
      verification_document_type: ObjectMapper.serialize(data),
      document: {
        id: data.id,
        name: source.name,
        attachment_type: source.type,
        mime_type: source.type,
        link: source.uri,
      },
      is_local_document: true,
    };
    const { localDocuments } = this.state;
    this.setState({
      localDocuments: [...localDocuments, ObjectMapper.deserialize(ExistingVerificationDocuments, imageObject)],
    });
  };

  public updateState = (data: IDocsProps): void => {
    const { existingDocuments, localDocuments, clonedDocuments, key } = data;
    this.setState((prevState) => ({
      ...prevState,
      ...(existingDocuments && { existingDocuments }),
      ...(localDocuments && { localDocuments }),
      ...(clonedDocuments && key && { [key]: clonedDocuments }),
    }));
  };

  // HANDLERS END

  // API'S START

  public postPropertyVerificationDocuments = async (): Promise<void> => {
    const { propertyId, updateStep, t, lastVisitedStep, typeOfPlan } = this.props;
    const { localDocuments, existingDocuments } = this.state;

    const updateAssetPayload: IUpdateAssetParams = {
      last_visited_step: {
        ...lastVisitedStep,
        listing: {
          ...lastVisitedStep.listing,
          type: typeOfPlan,
          is_verification_done: true,
        },
      },
    };

    if (localDocuments.length === 0) {
      await AssetRepository.updateAsset(propertyId, updateAssetPayload);
      updateStep();
      return;
    }
    const formData = new FormData();
    localDocuments.forEach((document: ExistingVerificationDocuments) => {
      formData.append('files[]', {
        // @ts-ignore
        name: document.document.name,
        uri: document.document.link,
        // @ts-ignore
        type: document.document.mimeType,
      });
    });

    this.setState({ isLoading: true });

    try {
      const response = await AttachmentService.uploadImage(formData, AttachmentType.ASSET_VERIFICATION);

      const { data } = response;
      const postRequestBody: IPostVerificationDocuments[] = [];

      localDocuments.forEach((document: ExistingVerificationDocuments, index: number) => {
        postRequestBody.push({
          verification_document_type_id: document.verificationDocumentType.id,
          document_id: data[index].id,
        });
      });

      existingDocuments.forEach((document: ExistingVerificationDocuments) => {
        if (!document.id) {
          postRequestBody.push({
            verification_document_type_id: document.verificationDocumentType.id,
            document_id: document.document.id,
          });
        }
      });
      this.setState({ isLoading: false });

      await AssetRepository.postVerificationDocuments(propertyId, postRequestBody);
      await ListingService.getExistingDocuments(propertyId, this.updateState);
      await AssetRepository.updateAsset(propertyId, updateAssetPayload);
      updateStep();
    } catch (e) {
      this.setState({ isLoading: false });
      if (e === AttachmentError.UPLOAD_IMAGE_ERROR) {
        AlertHelper.error({ message: t('common:fileCorrupt') });
      }
    }
  };

  // API'S END
}

export default withTranslation(LocaleConstants.namespacesKey.property)(PropertyVerification);

const styles = StyleSheet.create({
  buttonStyle: {
    flex: 0,
    marginVertical: 20,
  },
});
