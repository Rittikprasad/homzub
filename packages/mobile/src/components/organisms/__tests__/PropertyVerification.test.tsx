import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyVerification } from '@homzhub/mobile/src/components/organisms/PropertyVerification';
import { PropertyVerificationTypes } from '@homzhub/common/src/mocks/PropertyVerification';

let props: any;
let wrapper: ShallowWrapper;

describe('PropertyVerification', () => {
  const createTestProps = (testProps: any): object => ({
    navigateToPropertyHelper: jest.fn(),
    typeOfFlow: 'IDENTITY',
    updateStep: jest.fn(),
    propertyId: 1,
    setLoading: jest.fn(),
    ...testProps,
  });

  props = createTestProps({});

  it('should match snapshot', () => {
    wrapper = shallow(<PropertyVerification {...props} t={(key: string): string => key} />);
    wrapper.setState({ verificationTypes: PropertyVerificationTypes });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
