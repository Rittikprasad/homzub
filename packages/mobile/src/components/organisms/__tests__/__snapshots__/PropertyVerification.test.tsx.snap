// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyVerification should match snapshot 1`] = `
<Fragment>
  <Component
    deleteDocument={[Function]}
    existingDocuments={Array []}
    handleTypes={[Function]}
    handleUpload={[Function]}
    localDocuments={Array []}
  />
  <Button
    containerStyle={
      Object {
        "flex": 0,
        "marginVertical": 20,
      }
    }
    disabled={true}
    onPress={[Function]}
    title="common:continue"
    type="primary"
  />
</Fragment>
`;
